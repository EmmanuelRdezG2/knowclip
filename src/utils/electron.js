import electron from 'electron'

const {
  remote: { dialog },
} = electron

export const showSaveDialog = (name, extensions) =>
  new Promise((res, rej) => {
    try {
      dialog.showSaveDialog({ filters: [{ name, extensions }] }, filename => {
        res(filename)
      })
    } catch (err) {
      rej(err)
    }
  })
