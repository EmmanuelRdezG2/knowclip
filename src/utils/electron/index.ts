import helpers from './helpers'
import spectronMocks from '../../test/spectronMocks'

const { mocked, resetMocks: _resetMocks } = spectronMocks(
  'electron-helper',
  helpers
)

const {
  showSaveDialog,
  showOpenDialog,
  showOpenDirectoryDialog,
  showOpenDirectoriesDialog,
  openInBrowser,
  showMessageBox,
} =
  process.env.REACT_APP_SPECTRON || process.env.NODE_ENV === 'test'
    ? mocked
    : helpers

export {
  showSaveDialog,
  showOpenDialog,
  showOpenDirectoryDialog,
  showOpenDirectoriesDialog,
  openInBrowser,
  showMessageBox,
}
