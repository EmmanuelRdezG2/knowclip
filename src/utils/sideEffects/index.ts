import sideEffects from './module'
import spectronMocks from '../../test/spectronMocks'

const {
  mocked,
  mockFunctions: mockSideEffects,
  logMocks: logSideEffectsMocks,
} = spectronMocks('side-effect', sideEffects)

export { mockSideEffects, logSideEffectsMocks }

const { uuid, nowUtcTimestamp } =
  process.env.REACT_APP_CHROMEDRIVER || process.env.NODE_ENV === 'test'
    ? mocked
    : sideEffects

export { uuid, nowUtcTimestamp }
