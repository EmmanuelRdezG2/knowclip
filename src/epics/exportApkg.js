import { flatMap, tap, map } from 'rxjs/operators'
import { ofType, combineEpics } from 'redux-observable'
import { of, empty } from 'rxjs'
import * as r from '../redux'
import { join } from 'path'
import fs from 'fs'
import Exporter from 'anki-apkg-export-multi-field/dist/exporter'
import createTemplate from 'anki-apkg-export-multi-field/dist/template'
import tempy from 'tempy'
import { showSaveDialog } from '../utils/electron'
import { getApkgExportData } from '../utils/prepareExport'
import clipAudio from '../utils/clipAudio'
import { promisify } from 'util'

const wait = ms => new Promise(res => setTimeout(res, ms))

const writeFile = promisify(fs.writeFile)
const readFile = promisify(fs.readFile)

const exportApkgFailure = (action$, state$) =>
  action$.pipe(
    ofType('EXPORT_APKG_FAILURE'),
    tap(() => (document.body.style.cursor = 'default')),
    flatMap(({ errorMessage }) =>
      errorMessage
        ? of(
            r.simpleMessageSnackbar(
              `There was a problem making clips: ${errorMessage}`
            )
          )
        : empty()
    )
  )
const exportApkgSuccess = (action$, state$) =>
  action$.pipe(
    ofType('EXPORT_APKG_SUCCESS'),
    tap(() => (document.body.style.cursor = 'default')),
    map(({ successMessage }) => r.simpleMessageSnackbar(successMessage))
  )

const exportApkg = (action$, state$) =>
  action$.pipe(
    ofType('EXPORT_APKG_REQUEST'),
    flatMap(async ({ clipIds }) => {
      const directory = tempy.directory()

      document.body.style.cursor = 'progress'
      await wait(10)

      const outputFilePath = await showSaveDialog('Anki deck file', ['apkg'])

      if (!outputFilePath) return r.exportApkgFailure()

      try {
        const currentProjectMetadata = r.getCurrentProject(state$.value)
        if (!currentProjectMetadata)
          return r.exportApkgFailure('Could not find project')
        const noteType = r.getCurrentNoteType(state$.value)
        if (!noteType) return r.exportApkgFailure('Could not find note type')

        const exportData = getApkgExportData(
          state$.value,
          currentProjectMetadata,
          clipIds
        )

        const apkg = new Exporter(exportData.deckName, {
          sql: window.SQL,
          template: createTemplate(exportData.template),
        })

        await Promise.all(
          exportData.clips.map(async clipSpecs => {
            const {
              outputFilename,
              sourceFilePath,
              startTime,
              endTime,
            } = clipSpecs
            const clipOutputFilePath = join(directory, outputFilename)
            await clipAudio(
              sourceFilePath,
              startTime,
              endTime,
              clipOutputFilePath
            )
            apkg.addMedia(outputFilename, await readFile(clipOutputFilePath)) // async this?
          })
        )

        exportData.clips.forEach(
          ({ flashcardSpecs: { fields, ...restSpecs } }) => {
            apkg.addCard(fields, restSpecs)
          }
        )

        await apkg
          .save({
            type: 'nodebuffer',
            base64: false,
            compression: 'DEFLATE',
          })
          .then(async zip => {
            const promise = writeFile(outputFilePath, zip, 'binary')
            console.log(`Package has been generated: ${outputFilePath}`)
            return promise
          })

        return r.exportApkgSuccess('Flashcards made in ' + outputFilePath)
      } catch (err) {
        console.error(err)
        return r.exportApkgFailure(err)
      }
    })
  )

export default combineEpics(exportApkg, exportApkgSuccess, exportApkgFailure)
