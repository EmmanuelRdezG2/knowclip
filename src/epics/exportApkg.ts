import {
  flatMap,
  tap,
  map,
  catchError,
  mergeAll,
  startWith,
  endWith,
  concat,
  concatMap,
  filter,
  takeUntil,
  take,
  ignoreElements,
  switchMap,
} from 'rxjs/operators'
import { ofType, combineEpics, ActionsObservable } from 'redux-observable'
import { of, empty, defer, from, EMPTY } from 'rxjs'
import r from '../redux'
import tempy from 'tempy'
import * as anki from '@silvestre/mkanki'
import sql from 'better-sqlite3'
import { getApkgExportData } from '../utils/prepareExport'
import { showSaveDialog } from '../utils/electron'
import { areSameFile } from '../utils/files'
import { afterUpdates } from '../utils/afterUpdates'
import A from '../types/ActionType'
import { processNoteMedia, AnkiNoteMedia } from '../utils/ankiNote'
import { Database } from 'better-sqlite3'
import * as fs from 'fs'
import archiver from 'archiver'

const exportApkgFailure: AppEpic = (action$) =>
  action$.pipe(
    ofType<Action, ExportApkgFailure>(A.exportApkgFailure),
    tap(() => (document.body.style.cursor = 'default')),
    flatMap(({ errorMessage }) =>
      errorMessage
        ? of(
            r.simpleMessageSnackbar(
              `There was a problem making clips: ${errorMessage}`
            )
          )
        : empty()
    )
  )
const exportApkgSuccess: AppEpic = (action$) =>
  action$.pipe(
    ofType<Action, ExportApkgSuccess>(A.exportApkgSuccess),
    tap(() => (document.body.style.cursor = 'default')),
    map(({ successMessage }) => r.simpleMessageSnackbar(successMessage))
  )

const exportApkg: AppEpic = (action$, state$) =>
  action$.pipe(
    ofType<Action, ExportApkgRequest>(A.exportApkgRequest),
    switchMap((exportApkgRequest) => {
      const { mediaFileIdsToClipIds } = exportApkgRequest
      const directory = tempy.directory()

      const currentProject = r.getCurrentProject(state$.value)
      if (!currentProject)
        return of(r.exportApkgFailure('Could not find project'))

      const exportData = getApkgExportData(
        state$.value,
        currentProject,
        mediaFileIdsToClipIds
      )

      if ('missingMediaFiles' in exportData) {
        return getMissingMedia(
          exportData.missingMediaFiles,
          action$,
          exportApkgRequest
        )
      }

      return makeApkg(exportData, directory)
    })
  )

function makeApkg(exportData: ApkgExportData, directory: string) {
  return from(showSaveDialog('Anki APKG file', ['apkg'])).pipe(
    filter((path): path is string => Boolean(path)),
    flatMap((outputFilePath) => {
      document.body.style.cursor = 'progress'
      const pkg = new anki.Package()
      const deck = new anki.Deck(exportData.projectId, exportData.deckName)
      const noteModel = new anki.Model(exportData.noteModel)
      const clozeNoteModel = new anki.ClozeModel(exportData.clozeNoteModel)

      let processed = 0

      const processClipsObservables = exportData.clips.map(
        (clipSpecs: ClipSpecs) => {
          registerClip(deck, noteModel, clozeNoteModel, clipSpecs)

          return defer(async () => {
            const noteMediaResult = await processNoteMedia(clipSpecs, directory)
            if (noteMediaResult.errors)
              throw new Error(noteMediaResult.errors.join('; '))

            const noteMedia = noteMediaResult.value
            await addNoteMedia(pkg, noteMedia)

            const number = ++processed
            return r.setProgress({
              percentage: (number / exportData.clips.length) * 100,
              message: `${number} clips out of ${exportData.clips.length} processed`,
            })
          })
        }
      )
      return of(
        r.setProgress({
          percentage: 0,
          message: 'Processing clips...',
        })
      ).pipe(
        concat(from(processClipsObservables).pipe(mergeAll(20))),
        concat(
          of(
            r.setProgress({
              percentage: 100,
              message: 'Almost done! Saving .apkg file...',
            })
          ).pipe(
            concatMap(() => {
              pkg.addDeck(deck)
              const tmpFilename = tempy.file()
              return defer(async () => {
                const archive = archiver('zip')

                return new Promise((res, rej) => {
                  archive.on('error', (err) => {
                    console.error(`Problem with archive!`)
                    console.error(err)
                    rej(err)
                  })

                  archive.on('close', res)
                  archive.on('end', res)
                  archive.on('finish', res)

                  writeToFile(pkg, outputFilePath, {
                    db: sql(tmpFilename),
                    tmpFilename,
                    archive,
                  })
                })
              }).pipe(
                map(() =>
                  r.exportApkgSuccess('Flashcards made in ' + outputFilePath)
                ),
                endWith(r.setProgress(null))
              )
            })
          )
        )
      )
    }),
    catchError((err) => {
      console.error(err)
      return from([
        r.exportApkgFailure(err.message || err.toString()),
        r.setProgress(null),
      ])
    })
  )
}

function registerClip(
  deck: any,
  noteModel: any,
  clozeNoteModel: any,
  clipSpecs: ClipSpecs
) {
  const {
    flashcardSpecs: { fields, tags, id, clozeDeletions },
  } = clipSpecs

  const note = { fields, guid: id, tags }
  const clozeNote = clozeDeletions
    ? {
        fields: [clozeDeletions, ...fields.slice(1)],
        guid: `${id}__CLOZE`,
        tags,
      }
    : null

  deck.addNote(noteModel.note(note.fields, note.guid, note.tags))
  if (clozeNote)
    deck.addNote(
      clozeNoteModel.note(clozeNote.fields, clozeNote.guid, clozeNote.tags)
    )
}

async function addNoteMedia(pkg: any, noteMedia: AnkiNoteMedia) {
  const { soundData, imageData } = noteMedia
  pkg.addMedia(await soundData.data(), soundData.fileName)
  if (imageData) pkg.addMedia(await imageData.data(), imageData.fileName)
}

function getMissingMedia(
  missingMediaFiles: Array<MediaFile>,
  action$: ActionsObservable<Action>,
  exportApkgRequest: ExportApkgRequest
) {
  const missingMediaFileIds = missingMediaFiles.map((file) => file.id)
  const openMissingMediaFailure = action$.pipe(
    ofType<Action, OpenFileFailure>('openFileFailure'),
    filter(
      (a) =>
        a.file.type === 'MediaFile' && missingMediaFileIds.includes(a.file.id)
    ),
    take(1)
  )
  return from(missingMediaFiles).pipe(
    concatMap((file) =>
      of(r.openFileRequest(file)).pipe(
        concat(
          action$.pipe(
            ofType<Action, OpenFileSuccess>('openFileSuccess'),
            filter((a) => areSameFile(file, a.validatedFile)),
            take(1),
            ignoreElements()
          )
        )
      )
    ),
    startWith(r.closeDialog()),
    takeUntil(openMissingMediaFailure),
    endWith(
      r.reviewAndExportDialog(
        exportApkgRequest.mediaOpenPrior,
        exportApkgRequest.mediaFileIdsToClipIds
      )
    )
  )
}

interface AnkiPackage {
  write(db: Database): void
  media: Array<{
    filename: string
    name: string
    data: Buffer
  }>
}
function writeToFile(
  ankiPackage: AnkiPackage,
  filename: string,
  {
    db,
    tmpFilename,
    archive,
  }: { db: Database; tmpFilename: string; archive: archiver.Archiver }
) {
  ankiPackage.write(db)
  db.close()
  const out = fs.createWriteStream(filename)
  archive.pipe(out)

  if (!fs.existsSync(tmpFilename)) throw new Error('Problem creating db')

  archive.file(tmpFilename, { name: 'collection.anki2' })
  const media_info: { [i: string]: string } = {}
  ankiPackage.media.forEach((m, i) => {
    if (m.filename != null) archive.file(m.filename, { name: i.toString() })
    else archive.append(m.data, { name: i.toString() })
    media_info[i] = m.name
  })
  archive.append(JSON.stringify(media_info), { name: 'media' })
  archive.finalize()
}

export default combineEpics(exportApkg, exportApkgSuccess, exportApkgFailure)
