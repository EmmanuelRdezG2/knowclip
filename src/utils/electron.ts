import electron, { FileFilter } from 'electron'

const {
  remote: { dialog, getCurrentWindow },
} = electron

const win = getCurrentWindow()
export const showSaveDialog = (
  name: string,
  extensions: Array<string>
): Promise<string | null> =>
  new Promise((res, rej) => {
    try {
      dialog.showSaveDialog(
        win,
        { filters: [{ name, extensions }] },
        filename => {
          res(filename)
        }
      )
    } catch (err) {
      rej(err)
    }
  })

export const showOpenDialog = (
  filters: Array<FileFilter> = [],
  multiSelections = false
): Promise<Array<string> | null> =>
  new Promise((res, rej) => {
    try {
      dialog.showOpenDialog(
        win,
        {
          properties: multiSelections
            ? ['openFile', 'multiSelections']
            : ['openFile'],
          filters,
        },
        filePaths => res(filePaths)
      )
    } catch (err) {
      rej(err)
    }
  })

export const showOpenDirectoryDialog = (showHiddenFiles = true) =>
  // ): Promise<Array<string | null> | null> =>
  new Promise((res, rej) => {
    try {
      dialog.showOpenDialog(
        win,
        {
          properties: showHiddenFiles
            ? ['openDirectory', 'showHiddenFiles']
            : ['openDirectory'],
        },
        directoryPaths => res(directoryPaths ? directoryPaths[0] : null)
      )
    } catch (err) {
      rej(err)
    }
  })
