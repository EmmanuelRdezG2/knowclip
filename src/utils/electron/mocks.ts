import helpers from './helpers'
import { ipcRenderer } from 'electron'

type Mocks = Partial<
  { -readonly [K in keyof typeof helpers]: typeof helpers[K] }
>

export const mocked: Mocks = {}
export const resetMocks = () => {
  for (const k in mocked) delete mocked[k as keyof typeof helpers]
}

window.document.addEventListener('DOMContentLoaded', () => {
  let args: { [functionName: string]: string[] } = {}

  ipcRenderer.on('mock', (event, functionName, returnValue) => {
    const returnValues = (args[functionName] = args[functionName] || [])
    returnValues.push(returnValue)
    // @ts-ignore
    mocked[functionName] = () => {
      if (returnValues.length) return returnValues.shift()

      // @ts-ignore
      return helpers[functionName]
    }
  })

  ipcRenderer.on('reset-mocks', () => {
    args = {}
    resetMocks()
  })
})

const showSaveDialog: typeof helpers['showSaveDialog'] = async (...args) => {
  return mocked.showSaveDialog ? await mocked.showSaveDialog(...args) : null
}
const showOpenDialog: typeof helpers['showOpenDialog'] = async (...args) =>
  mocked.showOpenDialog ? await mocked.showOpenDialog(...args) : null
const showOpenDirectoryDialog: typeof helpers['showOpenDirectoryDialog'] = async (
  ...args
) =>
  mocked.showOpenDirectoryDialog
    ? await mocked.showOpenDirectoryDialog(...args)
    : null
const openInBrowser: typeof helpers['openInBrowser'] = (...args) =>
  mocked.openInBrowser ? mocked.openInBrowser(...args) : null

const showMessageBox: typeof helpers['showMessageBox'] = async (...args) =>
  mocked.showMessageBox ? await mocked.showMessageBox(...args) : null

export default {
  showSaveDialog,
  showOpenDialog,
  showOpenDirectoryDialog,
  openInBrowser,
  showMessageBox,
}
