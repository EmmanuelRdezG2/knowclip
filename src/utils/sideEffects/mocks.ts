import sideEffects from './module'
import spectronMocks from '../../test/spectronMocks'

const {
  mockFunctions: mockSideEffects,
  logMocks: logSideEffectsMocks,
} = spectronMocks('side-effect', sideEffects)

export { mockSideEffects, logSideEffectsMocks }
