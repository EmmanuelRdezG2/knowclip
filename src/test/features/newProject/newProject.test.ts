import { Application } from 'spectron'
import {
  startApp,
  stopApp,
  TestSetup,
  TMP_DIRECTORY,
  FIXTURES_DIRECTORY,
} from '../../setup'
import { mkdirp, remove, existsSync, copy } from 'fs-extra'
import createNewProject from './createNewProject'
import changeProjectName from './changeProjectName'
import addFirstMediaToProject from './addFirstMediaToProject'
import makeSomeFlashcards from './makeSomeFlashcards'
import navigateBetweenClips from './navigateBetweenClips'
import addMoreMediaToProject from './addMoreMediaToProject'
import navigateBetweenMedia from './navigateBetweenMedia'
import exportApkg from './exportApkg'
import moveThroughoutMedia from './moveThroughoutMedia'
import saveProject from './saveProject'
import { mockSideEffects } from '../../../utils/sideEffects'
import { mockElectronHelpers } from '../../../utils/electron/mocks'

jest.setTimeout(60000)

// describe('create a deck from a previously opened project', () => {
// describe('create a deck from a shared project', () => {
describe('create a deck from a new project', () => {
  let context: { app: Application | null } = { app: null }
  let setup: TestSetup

  beforeAll(async () => {
    if (existsSync(TMP_DIRECTORY)) await remove(TMP_DIRECTORY)
    await mkdirp(TMP_DIRECTORY)
    await copy(FIXTURES_DIRECTORY, TMP_DIRECTORY)
    setup = await startApp(context, 'newProjectTest')

    await mockSideEffects(setup.app, {
      uuid: [
        'e36ca6de-4893-49f5-bca1-4410ace25d46',
        '535935a3-7d72-4238-87ab-1b7a413c1f71',
        '37bd5e91-89c5-491b-9308-533c7be7338a',
        'ae3239c8-90bd-4488-95e3-bde059632348',
        '7c348315-8f78-42fd-9be7-b05bc0aade9d',
        'e62c823d-93c2-4a5e-8181-37e3d540c3e3',
        'fa462524-bfeb-4434-a506-07858cc97151',
      ],
      nowUtcTimestamp: [
        '2020-01-20T11:42:45Z',
        '2020-01-20T11:42:45Z',
        '2020-01-20T11:42:45Z',
        '2020-01-20T11:43:24Z',
        '2020-01-20T11:43:27Z',
      ],
    })
  })

  test('create a new project', async () => {
    return createNewProject(setup)
  }) // should create file
  test('change project name', async () => changeProjectName(setup))
  test('add media to project', async () => addFirstMediaToProject(setup))
  test('create clips + flashcards', async () => makeSomeFlashcards(setup))
  // navigate to end of video and add one more flashcard
  test('navigate between clips', async () => navigateBetweenClips(setup))
  test('move throughout media file', async () => moveThroughoutMedia(setup))
  test('add more media to project', async () => addMoreMediaToProject(setup))
  // stretch clips
  test('navigate between media', async () => navigateBetweenMedia(setup))
  // test('create clips + flashcards', async () => makeTwoFlashcards(setup))
  // test('create flashcards from subtitles', async () => creatFlashcardsFromSubtitles(setup))
  test('exporting an .apkg', async () => exportApkg(setup))
  test('saving a project file', async () => saveProject(setup))
  // test project file matches example
  // and stuff was saved to local storage?

  afterAll(async () => {
    await mockElectronHelpers(setup.app, {
      showMessageBox: [
        Promise.resolve({
          response: 0,
          checkboxChecked: false,
        }),
      ],
    })
    await stopApp(context)
  })
})

// describe('second test', () => {
//   let context: { app: Application | null } = { app: null }
//   let setup: TestSetup

//   beforeAll(async () => {
//     if (existsSync(TMP_DIRECTORY)) await remove(TMP_DIRECTORY)
//     await mkdirp(TMP_DIRECTORY)
//     // await copy(join(__dirname, 'fixtures'), TMP_DIRECTORY)
//     setup = await startApp(context)
//   })

//   test('create a new project', async () => {
// return (setup))
//   test('change project name', async () => changeProjectName(setup))
//   test('add media to project', async () => addJapaneseMedia(setup))
//   test('create flashcards', async () => makeTwoFlashcards(setup))

//   afterAll(async () => {
//     await mockElectronHelpers(setup.app, {
//       showMessageBox: Promise.resolve({
//         response: 0,
//         checkboxChecked: false,
//       }),
//     })
//     await stopApp(context)
//   })
// })
