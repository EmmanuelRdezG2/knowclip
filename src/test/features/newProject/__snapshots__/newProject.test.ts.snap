// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create a deck from a new project saving a project file 1`] = `
Object {
  "clips": Array [
    Object {
      "end": 625,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Relaxing while eating bamboo grass is the best",
          "notes": "",
          "pronunciation": "sasa-o tabe-nágara nonbíri-suru-no-wa saikoo-da-naa",
          "transcription": "笹を食べながらのんびりするのは最高だなぁ",
        },
        "id": "ae3239c8-90bd-4488-95e3-bde059632348",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "ae3239c8-90bd-4488-95e3-bde059632348",
      "start": 402,
    },
    Object {
      "end": 888,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "This kid, lazing about again so early",
          "notes": "\\"Goro-goro\\" is the sound of something big rolling around.",
          "pronunciation": "mata kono ko-wa hirumá-kara góro-goro shite",
          "transcription": "またこの子は昼間からゴロゴロして",
        },
        "id": "7c348315-8f78-42fd-9be7-b05bc0aade9d",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "7c348315-8f78-42fd-9be7-b05bc0aade9d",
      "start": 756,
    },
    Object {
      "end": 4054,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "",
          "notes": "",
          "pronunciation": "",
          "transcription": "",
        },
        "id": "e62c823d-93c2-4a5e-8181-37e3d540c3e3",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "e62c823d-93c2-4a5e-8181-37e3d540c3e3",
      "start": 3756,
    },
  ],
  "id": "e36ca6de-4893-49f5-bca1-4410ace25d46",
  "lastOpened": "2020-01-20T11:42:45Z",
  "mediaFiles": Array [
    Object {
      "durationSeconds": 87.062,
      "flashcardFieldsToSubtitlesTracks": Object {},
      "format": "mov,mp4,m4a,3gp,3g2,mj2",
      "id": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "isVideo": true,
      "name": "polar_bear_cafe.mp4",
      "parentId": "e36ca6de-4893-49f5-bca1-4410ace25d46",
      "subtitles": Array [
        Object {
          "id": "37bd5e91-89c5-491b-9308-533c7be7338a",
          "streamIndex": 2,
          "type": "EmbeddedSubtitlesTrack",
        },
      ],
      "subtitlesTracksStreamIndexes": Array [
        2,
      ],
      "type": "MediaFile",
    },
    Object {
      "durationSeconds": 58.24,
      "flashcardFieldsToSubtitlesTracks": Object {},
      "format": "mov,mp4,m4a,3gp,3g2,mj2",
      "id": "fa462524-bfeb-4434-a506-07858cc97151",
      "isVideo": true,
      "name": "piggeldy_cat.mp4",
      "parentId": "e36ca6de-4893-49f5-bca1-4410ace25d46",
      "subtitles": Array [],
      "subtitlesTracksStreamIndexes": Array [],
      "type": "MediaFile",
    },
  ],
  "name": "My cool new project",
  "noteType": "Transliteration",
  "subtitles": Array [],
  "tags": Array [
    "pbc",
  ],
  "timestamp": "2020-01-20T11:43:27Z",
  "version": "4.1.0",
}
`;
